src/CLI/klish/patches/klish-2.1.4/plugins/clish/rest_cl.cpp

```diff
@@ -0,0 +1,373 @@
+    if (!ietf_err) {
+        syslog(LOG_DEBUG, "clish_restcl: No errors\r\n");
+        return 0;
+    }
+
+    cJSON *errors = cJSON_GetObjectItemCaseSensitive(ietf_err, "error");
+    if (!errors) {
+        syslog(LOG_DEBUG, "clish_restcl: No error\r\n");
+        return 0;
+    }
+
+    cJSON *error;
+    cJSON_ArrayForEach(error, errors) {
+        cJSON *err_msg = cJSON_GetObjectItemCaseSensitive(error, "error-message");
+        if (err_msg) {
+            lub_dump_printf("%% Error: %s\r\n", err_msg->valuestring);
+        }
+        return 1;
+    }
+
+    return 0;
+}
+
+
+std::string rest_token;
+
+CURL *curl =  NULL;
+
+static int rest_set_curl_headers(bool use_token) {
+    struct curl_slist* headerList = NULL;
+    headerList = curl_slist_append(headerList, "accept: application/yang-data+json");
+    headerList = curl_slist_append(headerList, "Content-Type: application/yang-data+json");
+
+    if (rest_token.size() && use_token) {
+        std::string auth_hdr = "Authorization: Bearer ";
+        auth_hdr += rest_token;
+        headerList = curl_slist_append(headerList, auth_hdr.c_str());
+    }
+
+    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headerList);
+
+    return 0;
+}
+
+std::string user_cert, user_key;
+
+void rest_update_user_info() {
+    char *cli_user = getenv("CLI_USER");
+    if (cli_user) {
+        struct passwd *p;
+        p=getpwnam(cli_user);
+        if (p) {
+            user_cert = p->pw_dir;
+            user_cert.append("/.cert/certificate.pem");
+
+            user_key = p->pw_dir;
+            user_key.append("/.cert/key.pem");
+            
+            setenv("USER_CERT_PATH", user_cert.c_str(), 1);
+            setenv("USER_KEY_PATH", user_key.c_str(), 1);
+            
+            syslog(LOG_DEBUG, "clish_restcl: [key:%s][cert:%s]\r\n",
+                    user_key.c_str(), user_cert.c_str());
+        }
+    }
+}
+
+static void rest_set_curl_cert_key() {
+    if (user_cert.size() && user_key.size()){
+        curl_easy_setopt(curl, CURLOPT_SSLCERT, user_cert.c_str());
+        curl_easy_setopt(curl, CURLOPT_SSLKEY, user_key.c_str());
+    }
+}
+
+static int _init_curl() {
+
+    char *root = getenv("REST_API_ROOT");
+
+    if (root) {
+        REST_API_ROOT.assign(root);
+    } else {
+        REST_API_ROOT.assign("https://localhost:8443");
+    }
+
+    curl_global_init(CURL_GLOBAL_ALL);
+
+    curl = curl_easy_init();
+    if (!curl) {
+        return 1;
+    }
+    
+    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
+    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
+
+    curl_easy_setopt(curl, CURLOPT_USERAGENT, "CLI");
+
+    curl_easy_setopt(curl, CURLOPT_READFUNCTION, read_callback);
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
+
+    return 0;
+}
+
+
+int rest_token_fetch(int *interval) {
+
+    CURLcode res;
+    std::string url("https://localhost:8443/authenticate");
```

Comment:
Please read root path from REST_API_ROOT environment variable.

Also, is it possible to disable this JWT authentication through any command line parameter or environment variable? We do run rest server and CLI in no-auth mode on our development servers & VMs. There will not be any CLI specific listener in that mode.

Reply:
Will look into providing a way to run CLI in no-auth mode.